pgjdbc/src/main/java/org/postgresql/core/AsciiStringInterner.java

```diff
@@ -0,0 +1,310 @@
+   * either an existing value will be returned, or the newly created {@code String} will be stored before being
+   * returned.
+   *
+   * <p>
+   * If non-ascii bytes are discovered, the <i>encoding</i> will be used to
+   * {@link Encoding#decode(byte[], int, int) decode} and that value will be returned (but not stored).
+   * </p>
+   *
+   * @param bytes The bytes of the String. Must not be {@code null}.
+   * @param offset Offset into <i>bytes</i> to start.
+   * @param length The number of bytes in <i>bytes</i> which are relevant.
+   * @param encoding To use if non-ascii bytes seen.
+   * @return Decoded {@code String} from <i>bytes</i>.
+   * @throws IOException If error decoding from <i>Encoding</i>.
+   */
+  public String getString(byte[] bytes, int offset, int length, Encoding encoding) throws IOException {
+    final int hash = hashKey(bytes, offset, length);
+    // 0 indicates the presence of a non-ascii character - defer to encoding to create the string
+    if (hash == 0) {
+      return encoding.decode(bytes, offset, length);
+    }
+    cleanQueue();
+    // create a TempKey with the byte[] given
+    final TempKey tempKey = new TempKey(hash, bytes, offset, length);
+    SoftReference<String> ref = cache.get(tempKey);
+    if (ref != null) {
+      final String val = ref.get();
+      if (val != null) {
+        return val;
+      }
+    }
+    // in order to insert we need to create a "real" key with copy of bytes that will not be changed
+    final byte[] copy = Arrays.copyOfRange(bytes, offset, offset + length);
+    final Key key = new Key(copy, hash);
+    final String value = new String(copy, StandardCharsets.US_ASCII);
+
+    // handle case where a concurrent thread has populated the map or existing value has cleared reference
+    ref = cache.compute(key, (k,v) -> {
+      if (v == null) {
+        return new StringReference(key, value);
+      }
+      final String val = v.get();
+      return val != null ? v : new StringReference(key, value);
+    });
+
+    return castNonNull(ref.get());
+  }
+
+  /**
+   * Process any entries in {@link #refQueue} to purge from the {@link #cache}.
+   * @see StringReference#dispose()
+   */
+  private void cleanQueue() {
+    Reference<?> ref;
+    while ((ref = refQueue.poll()) != null) {
+      ((StringReference)ref).dispose();
+    }
+  }
+
+  /**
+   * Generates a hash value for the relevant entries in <i>bytes</i> as long as all values are ascii ({@code >= 0}).
+   * @return hash code for relevant bytes, or {@code 0} if non-ascii bytes present.
+   */
+  private static int hashKey(byte[] bytes, int offset, int length) {
+    int result = 1;
+    for (int i = offset, j = offset + length; i < j; ++i) {
+      final byte b = bytes[i];
+      // bytes are signed values. all ascii values are positive
+      if (b < 0) {
+        return 0;
+      }
+      result = 31 * result + b;
```

Comment:
If the string is long / odd enough that the hash actually hits zero, it just won't get interned right?

Reply:
correct

Reply:
Correct